### PR TITLE
BCDA-3849 - Add middleware component to check if ACO is blacklisted

### PR DESCRIPTION
**Pull request by mtrang1263**
### Fixes [BCDA-3849](https://jira.cms.gov/browse/BCDA-3849)

To support blacklisting ACOs, we need to be able to check and reject requests made by a blacklisted ACO.

### Change Details

1. Add new auth middleware component that checks to see if the ACO is blacklisted. If it is, we return a 403 Unauthorized response.
2. Add `commonAuth` variable to capture all of the auth checks that need to occur. This should help us avoid forgetting a particular check.
3. Update our init script to add new blacklisted ACO.
4. Update postman tests to verify that requests made by a blacklisted ACO fail.

### Security Implications

- [ ] new software dependencies
- [ ] security controls or supporting software altered
- [ ] new data stored or transmitted
- [ ] security checklist is completed for this change
- [ ] requires more information or team discussion to evaluate security implications
- [x] no PHI/PII is affected by this change

### Acceptance Validation
CI passes.

We have [another ticket ](https://jira.cms.gov/browse/BCDA-3967)to add smoke test verification

### CODE REVIEW

**Comment by msnook** [comment]
Is it true that we can't merge this PR until [this ticket](https://jira.cms.gov/browse/BCDA-3967) is complete?

File: test/smoke_test/smoke_test.sh [hunk 1]
```diff
@@ -33,6 +33,9 @@ do
         testFile=bulk_data_requests_lite.sh
         if [ ${CMS_ID} = "A9996" ]; then
                 testFile=bulk_data_requests.sh
+        elif [ ${CMS_ID} = "A9997" ]; then
+                docker-compose exec -T db sh -c 'psql "postgres://postgres:toor@db:5432/bcda?sslmode=disable" -c "UPDATE acos SET blacklisted=true WHERE cms_id = '"'"${CMS_ID}"'"'"'
```
Comment (msnook) [inline test/smoke_test/smoke_test.sh:37]: Why don't we use the new CLI instead?

File: bcda/auth/middleware.go [hunk 2]
```diff
@@ -116,6 +120,27 @@ func RequireTokenAuth(next http.Handler) http.Handler {
...
+		if !ok {
+			log.Error()
+			oo := responseutils.CreateOpOutcome(responseutils.Error, responseutils.Exception, responseutils.Not_found, "AuthData not found")
+			responseutils.WriteError(oo, w, http.StatusNotFound)
+			return
+		}
+
+		if ad.Blacklisted {
+			oo := responseutils.CreateOpOutcome(responseutils.Error, responseutils.Exception, responseutils.RequestErr,
+				fmt.Sprintf("ACO (CMS_ID: %s) is blacklisted", ad.CMSID))
```
Comment (msnook) [inline bcda/auth/middleware.go:136]: Do we want to use this term ("blacklisted") in a public facing error message?  Maybe we should just say something about being unauthorized?

File: bcda/auth/middleware.go [hunk 3]
```diff
@@ -116,6 +120,27 @@ func RequireTokenAuth(next http.Handler) http.Handler {
...
+		ad, ok := r.Context().Value(AuthDataContextKey).(AuthData)
+		if !ok {
+			log.Error()
+			oo := responseutils.CreateOpOutcome(responseutils.Error, responseutils.Exception, responseutils.Not_found, "AuthData not found")
+			responseutils.WriteError(oo, w, http.StatusNotFound)
+			return
+		}
+
+		if ad.Blacklisted {
+			oo := responseutils.CreateOpOutcome(responseutils.Error, responseutils.Exception, responseutils.RequestErr,
```
Comment (msnook) [inline bcda/auth/middleware.go:135]: Should we be more granular about the issue being an authorization exception?

File: test/smoke_test/smoke_test.sh (hunk 1 above)
Comment (mtrang1263) [reply test/smoke_test/smoke_test.sh:37]: I wanted to decouple the two PRs, but I think I can merge the CLI one first and then update the script to use the command.

**Comment by mtrang1263** [comment]
> Is it true that we can't merge this PR until [this ticket](https://jira.cms.gov/browse/BCDA-3967) is complete?

I don't think so. I ended up pulling out the smoke tests into a separate ticket to give us an opportunity to test the tests on all the environments. So we should have the functionality in place before we create the smoke tests. If it's reversed, we may encounter issues in prod (related to the smoke tests) that won't be caught until we're in the release window.

In terms of correctness, I believe that the tests implemented in this PR should give us enough confidence. The smoke tests on the various environments should be used to detect regressions later down the road.

File: test/smoke_test/smoke_test.sh (hunk 1 above)
Comment (mtrang1263) [reply test/smoke_test/smoke_test.sh]: This is no longer needed since the negative tests are being verified in the postman test suite.

**Comment by msnook** [comment]
> > Is it true that we can't merge this PR until [this ticket](https://jira.cms.gov/browse/BCDA-3967) is complete?
> 
> I don't think so. I ended up pulling out the smoke tests into a separate ticket to give us an opportunity to test the tests on all the environments. So we should have the functionality in place before we create the smoke tests. If it's reversed, we may encounter issues in prod (related to the smoke tests) that won't be caught until we're in the release window.
> 
> In terms of correctness, I believe that the tests implemented in this PR should give us enough confidence. The smoke tests on the various environments should be used to detect regressions later down the road.

But what about the Postman tests?  Won't they fail due to the way they are invoked currently?  I thought they would need to be invoked differently from `bcda-ops` when run against the AWS envs.  Isn't that the work to be done in BCDA-3967?

**Comment by mtrang1263** [comment]
> But what about the Postman tests? Won't they fail due to the way they are invoked currently? I thought they would need to be invoked differently from `bcda-ops` when run against the AWS envs. Isn't that the work to be done in BCDA-3967?

The postman tests have been updated to account for those environments. Each of the new blacklist tests have a pre-check that sees if we've passed in creds associated with a blacklisted ACO.

```
if (!pm.globals.get("blacklistedClientId") || pm.globals.get("blacklistedClientSecret")) {
    console.log("Blacklist test skipped due to creds not set.")
    return
}
```

An example is here:https://github.com/CMSgov/bcda-app/pull/582/files#diff-842f2e0111020a6f5af8ab0ad66d148de552df214ba184469f9587155f9d5337R2970

**Comment by msnook** [comment]
> > But what about the Postman tests? Won't they fail due to the way they are invoked currently? I thought they would need to be invoked differently from `bcda-ops` when run against the AWS envs. Isn't that the work to be done in BCDA-3967?
> 
> The postman tests have been updated to account for those environments. Each of the new blacklist tests have a pre-check that sees if we've passed in creds associated with a blacklisted ACO.
> 
> ```
> if (!pm.globals.get("blacklistedClientId") || !pm.globals.get("blacklistedClientSecret")) {
>     console.log("Blacklist test skipped due to creds not set.")
>     return
> }
> ```
> 
> An example is here:https://github.com/CMSgov/bcda-app/pull/582/files#diff-842f2e0111020a6f5af8ab0ad66d148de552df214ba184469f9587155f9d5337R2970
> 
> Edit: Link might not work, but if you see Line 2970 in the postman tests, we shortcircuit if we do not supply the expected credentials.

Heh, I didn't notice that.  I see it now... makes sense!

**Review by msnook — APPROVED** [review]
👍

**Comment by mtrang1263** [comment]
> > But what about the Postman tests? Won't they fail due to the way they are invoked currently? I thought they would need to be invoked differently from `bcda-ops` when run against the AWS envs. Isn't that the work to be done in BCDA-3967?
> 
> The postman tests have been updated to account for those environments. Each of the new blacklist tests have a pre-check that sees if we've passed in creds associated with a blacklisted ACO.
> 
> ```
> if (!pm.globals.get("blacklistedClientId") || !pm.globals.get("blacklistedClientSecret")) {
>     console.log("Blacklist test skipped due to creds not set.")
>     return
> }
> ```
> 
> An example is here:https://github.com/CMSgov/bcda-app/pull/582/files#diff-842f2e0111020a6f5af8ab0ad66d148de552df214ba184469f9587155f9d5337R2970
> 
> Edit: Link might not work, but if you see Line 2970 in the postman tests, we shortcircuit if we do not supply the expected credentials.

Here's a successful smoke test run on [dev](https://bcda-ci.adhocteam.us/blue/organizations/jenkins/BCDA%20-%20Test%20-%20Integration%20Smoke/detail/BCDA%20-%20Test%20-%20Integration%20Smoke/2558/pipeline/189) with this branch.

```

→ Get Blacklisted auth token

  POST https://dev.bcda.cms.gov/auth/token [401 Unauthorized, 315B, 108ms]

  ┌

  │ 'Blacklist test skipped due to creds not set.'

  └



→ Start Blacklisted Patient export

  GET https://dev.bcda.cms.gov/api/v1/Patient/$export [401 Unauthorized, 490B, 209ms]

  ┌

  │ 'Blacklist test skipped due to creds not set.'

  └



→ Start Blacklisted Group export

  GET https://dev.bcda.cms.gov/api/v1/Group/all/$export [401 Unauthorized, 490B, 160ms]

  ┌

  │ 'Blacklist test skipped due to creds not set.'

  └



→ Start Blacklisted Job retrieval

  GET https://dev.bcda.cms.gov/api/v1/jobs/1 [401 Unauthorized, 490B, 163ms]

  ┌

  │ 'Blacklist test skipped due to creds not set.'

  └



→ Start Blacklisted Data Retrieval

  GET https://dev.bcda.cms.gov/data/test/test.ndjson [401 Unauthorized, 490B, 199ms]

  ┌

  │ 'Blacklist test skipped due to creds not set.'
```

**Pull request merged**